### ISSUE TITLE
E2E conformance bypasses Kernel Registry/Dispatch by calling Reference CPU kernel functions directly

### ISSUE THREAD

**Issue by astorise** [astorise/Magnetar#27]
## Priority

**P0 — blocks `v0.1` E2E conformance**

## Context

The accepted E2E contract requires the first local inference conformance path to prove the complete Runtime architecture:

```text
client
  -> Runtime Inference API
  -> Model Loading
  -> Model Instance
  -> Tokenizer
  -> Generation
  -> Model Component
  -> Execution Graph
  -> Operators
  -> Kernel Registry
  -> Kernel Dispatch
  -> Reference CPU Provider
  -> Sampling
  -> output
```

The current E2E test performs real numerical computation, which is good, but it does not yet prove this complete path.

## Problem

`e2e_forward()` directly calls Reference CPU implementation functions such as:

```text
embedding_lookup
rmsnorm
matmul
rope
attention
softmax_rows
silu
mul
residual_add
```

The E2E suite separately validates that Kernel Registry can select a Reference CPU kernel, but the actual inference forward pass does not execute through that selected registry/dispatch path.

The effective structure is therefore:

```text
E2E
 ├── test Kernel Registry separately
 │
 └── call Reference CPU functions directly for inference
```

instead of:

```text
E2E
  -> Runtime
     -> Kernel Registry
        -> Kernel Dispatch
           -> Reference CPU Provider
```

The `no-shortcut` assertion also currently relies in part on caller-supplied boolean evidence rather than evidence produced by the actual execution trace.

## Why this matters

The current test proves useful things:

- the fixture is numerical rather than canned output;
- Reference CPU functions produce usable results;
- generation/sampling orchestration works;
- Kernel Registry works in isolation.

But it does **not** yet prove that the actual inference path cannot bypass the registry and dispatch layers.

That is exactly the invariant the `define-end-to-end-local-inference-conformance` change was intended to close.

## Required change

Refactor E2E conformance so its success path enters only through the public/approved Runtime Inference API.

The E2E fixture should not directly invoke:

```text
ReferenceCpuProvider kernel functions
Kernel implementations
KernelRegistry selection APIs
e2e_forward implementation shortcuts
```

The Runtime should execute those internally.

## Replace boolean proof with execution evidence

Avoid checks shaped like:

```rust
validate_e2e_no_shortcuts(
    used_kernel_registry: true,
    ...
)
```

where the caller can simply claim that a layer was used.

Prefer evidence derived from Runtime execution, for example redacted observations such as:

```text
model-instance-ready
graph-validated
kernel-selected
kernel-dispatched
provider-executed
sampling-executed
generation-completed
```

Exact observation names are implementation-defined.

## Acceptance criteria

- [ ] E2E success path enters through Runtime Inference API.
- [ ] E2E does not call Reference CPU kernel functions directly.
- [ ] E2E does not instantiate/use Kernel Registry directly from the test success path.
- [ ] Model Component produces or participates in producing the prefill/decode graph.
- [ ] Runtime validates the graph.
- [ ] Kernel Registry selects Reference CPU kernels.
- [ ] Kernel Dispatch invokes Reference CPU Provider.
- [ ] Tensor/Memory contracts participate in execution.
- [ ] Sampling receives logits from the dispatched execution.
- [ ] Generated result remains deterministic for the fixture.
- [ ] No-shortcut validation derives from actual execution evidence.
- [ ] Removing/bypassing Kernel Registry makes the E2E test fail.
- [ ] Removing/bypassing Model Loading makes the E2E test fail.
- [ ] Direct Provider execution shortcut makes conformance fail.

## Failure fixtures

Add or preserve cases for:

- invalid Model Artifact;
- untrusted artifact;
- invalid graph;
- unsupported Operator;
- missing Kernel;
- Provider unavailable;
- Memory Manager rejection;
- invalid tensor shape;
- cancellation;
- session lifecycle failure.

## OpenSpec coverage

This is an implementation/conformance gap against existing accepted specs:

- `define-end-to-end-local-inference-conformance`
- `define-runtime-inference-api`
- `define-kernel-registry-and-dispatch`
- `define-reference-cpu-provider-and-kernel-baseline`

No new OpenSpec change is required.

## Depends on

Runtime-owned model execution issue.

## Blocks

`v0.1`

**Comment by astorise** [comment]
PR #33 now routes the E2E success path through Runtime generation and validates redacted execution observations for graph validation, kernel selection, kernel dispatch, provider execution, and tensor logits. I am leaving this open because the strict acceptance criteria also require removing the fixture forward helper shortcut from the success path; that deeper refactor is still visible in docs/release-v0.1-evidence.md.

**Comment by astorise** [comment]
Latest PR #33 status: review correctly identified that the fixture executor still computes logits through e2e_forward and does not yet execute a real KernelDispatch + ProviderExecution path for the forward pass. I am leaving #27 open rather than certifying this shortcut. The branch now fixes the independent executor-error observability issue, but the remaining #27 work is the deeper refactor to produce logits from actual Runtime dispatch/Provider execution instead of fixture helper calls.

**Comment by astorise** [comment]
Closed by PR #33 as deferred from the v0.1 baseline. The E2E/CLI paths no longer claim KernelDispatched, ProviderExecuted, or TensorLogitsProduced evidence for shortcut/placeholder logits; the release evidence now documents full KernelDispatch/ProviderExecution-backed logits as deferred rather than certifying the shortcut path.
